### PR TITLE
Cache kubeconfig used for kubectl commands

### PR DESCRIPTION
**Pull request by rmweir**
Cache and reuse kubeconfig

### CODE REVIEW

File: cmd/common.go [hunk 1]
```diff
@@ -139,6 +139,18 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
...
+// serverCurrent command to display the name of the current server in the local config
+func serverCurrent(ctx *cli.Context) (string, string, error) {
+	cf, err := loadConfig(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	serverName := cf.CurrentServer
+	URL := cf.Servers[serverName].URL
+	return serverName, URL, nil
```
Comment (nickgerace) [inline cmd/common.go:151]: nit:
```go
return serverName, cf.Servers[serverName].URL, nil
```

File: cmd/kubectl.go [hunk 2]
```diff
@@ -5,8 +5,19 @@ import (
...
+	"github.com/sirupsen/logrus"
 	"github.com/urfave/cli"
+	"gopkg.in/yaml.v2"
+
+	"github.com/rancher/norman/clientbase"
+	client "github.com/rancher/types/client/management/v3"
+)
+
+const (
+	kubeconfigpath = ".rancher/.cache/kubeconfig"
```
Comment (nickgerace) [inline cmd/kubectl.go:20]: Is a relative path always going to work?

File: cmd/server.go [hunk 3]
```diff
@@ -64,15 +64,13 @@ be displayed and one can be selected.
...
+func printServerCurrent(ctx *cli.Context) error {
+	serverName, url, err := serverCurrent(ctx)
 	if err != nil {
 		return err
 	}
 
-	serverName := cf.CurrentServer
-	URL := cf.Servers[serverName].URL
-	fmt.Printf("Name: %s URL: %s\n", serverName, URL)
+	fmt.Printf("Name: %s URL: %s\n", serverName, url)
```
Comment (nickgerace) [inline cmd/server.go:73]: nit: Should we add a second `\n` character before URL or some other form of separation?

File: cmd/kubectl.go [hunk 4]
```diff
@@ -37,34 +48,66 @@ func runKubectl(ctx *cli.Context) error {
...
 
-	_, err = tmpfile.Write([]byte(config.Config))
+	currentUser := t.UserID
+	currentCluster := currentRancherServer.FocusedCluster()
+	filename := fmt.Sprintf("%s-%s-kubeconfig", currentUser, currentCluster)
+	home, err := os.UserHomeDir()
 	if err != nil {
 		return err
 	}
+	kubeconfigFilePath := fmt.Sprintf("%s/%s/%s", home, kubeconfigpath, filename)
```
Comment (nickgerace) [inline cmd/kubectl.go:74]: nit: Using `path.Join` or something similar might be better here.
https://golang.org/pkg/path/#Join

**Review by nickgerace — APPROVED** [review]
Some small nits, but otherwise, LGTM

File: cmd/common.go [hunk 5]
```diff
@@ -139,6 +139,18 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
 	return writer.Err()
 }
 
+// serverCurrent command to display the name of the current server in the local config
+func serverCurrent(ctx *cli.Context) (string, string, error) {
```
Comment (aiyengar2) [inline cmd/common.go:143]: nit: I think this should be renamed to
```suggestion
func getCurrentServerFromConfig(ctx *cli.Context) (string, string, error) {
```

`getCurrentServer` or `currentServer` would be okay as well.

File: cmd/kubectl.go [hunk 6]
```diff
@@ -37,34 +48,66 @@ func runKubectl(ctx *cli.Context) error {
...
+	t, err := c.ManagementClient.Token.ByID(currentToken)
 	if err != nil {
 		return err
 	}
-	defer os.Remove(tmpfile.Name())
 
-	_, err = tmpfile.Write([]byte(config.Config))
+	currentUser := t.UserID
+	currentCluster := currentRancherServer.FocusedCluster()
+	filename := fmt.Sprintf("%s-%s-kubeconfig", currentUser, currentCluster)
```
Comment (aiyengar2) [inline cmd/kubectl.go:69]: Should this format be a constant at the top of the file?

File: cmd/kubectl.go [hunk 7]
```diff
@@ -37,34 +48,66 @@ func runKubectl(ctx *cli.Context) error {
...
+		}
+		kubeconfig = []byte(config.Config)
+		if err := writeKubeconfig(kubeconfigFilePath, kubeconfig); err != nil {
+			return err
+		}
+	}
+
 	cmd := exec.Command(path, ctx.Args()...)
-	cmd.Env = append(os.Environ(), "KUBECONFIG="+tmpfile.Name())
+	cmd.Env = append(os.Environ(), "KUBECONFIG="+kubeconfigFilePath)
```
Comment (aiyengar2) [inline cmd/kubectl.go:110]: Should we be using [`os.Setenv`](https://golang.org/pkg/os/#Setenv) here?

File: cmd/kubectl.go [hunk 8]
```diff
@@ -74,3 +117,75 @@ func runKubectl(ctx *cli.Context) error {
...
+}
+
+func extractKubeconfigTokenID(kubeconfig []byte) (string, error) {
+	type Token struct {
+		Users []struct {
+			User struct {
+				T string `yaml:"token"`
+			} `yaml:"user"`
+		} `json:"users,omitempty" yaml:"users,omitempty"`
+	}
```
Comment (aiyengar2) [inline cmd/kubectl.go:164]: Should this be declared outside of the function?

File: cmd/server.go [hunk 9]
```diff
@@ -64,15 +64,13 @@ be displayed and one can be selected.
 }
 
 // serverCurrent command to display the name of the current server in the local config
-func serverCurrent(ctx *cli.Context) error {
-	cf, err := loadConfig(ctx)
+func printServerCurrent(ctx *cli.Context) error {
```
Comment (aiyengar2) [inline cmd/server.go:67]:
```suggestion
func printCurrentServer(ctx *cli.Context) error {
```

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: cmd/kubectl.go (hunk 2 above)
Comment (rmweir) [reply cmd/kubectl.go:20]: changed to use HOME env var

File: cmd/server.go (hunk 3 above)
Comment (rmweir) [reply cmd/server.go:73]: I don't think so. I was really just extracting some logic to the common.go, but I don't see the need to change printing format that is unrelated to the ticket since it's not objectively worse or anything.

File: cmd/kubectl.go (hunk 4 above)
Comment (rmweir) [reply cmd/kubectl.go:74]: Changed but is there any benefit of using that instead when the number of strings is already known apart from it being in that path package?

File: cmd/kubectl.go (hunk 6 above)
Comment (rmweir) [reply cmd/kubectl.go:69]: The format is only used once so it could be but is there a benefit?

File: cmd/kubectl.go (hunk 7 above)
Comment (rmweir) [reply cmd/kubectl.go:110]: I think this way may be better since the var is only needed for the single command and this contains it.

File: cmd/kubectl.go (hunk 4 above)
Comment (aiyengar2) [reply cmd/kubectl.go:74]: I think it's a better practice because it takes care of weird edge cases (e.g. if `home`, `kubeconfigpath`, or `filename` have trailing slashes or something similar).

File: cmd/kubectl.go (hunk 6 above)
Comment (aiyengar2) [reply cmd/kubectl.go:69]: Hmm not sure. I like to keep these as constants on the top even if it's used once since I don't like magic strings, but I'm fine with either decision!

File: cmd/kubectl.go (hunk 7 above)
Comment (aiyengar2) [reply cmd/kubectl.go:110]: Oh sorry you're right. I didn't realize this was just being set for this command.

File: cmd/kubectl.go (hunk 8 above)
Comment (rmweir) [reply cmd/kubectl.go:164]: I considered this as well, I decided to contain it within the function because it's strictly being used to unmarshal, and only to extract the token at that so idk if it would ever have a user outside of the function since it is incomplete.

**Comment by rmweir** [comment]
@aiyengar2 @nickgerace respond to and or addressed feedback. Fixed a few additional things I noticed as well.

File: cmd/kubectl.go (hunk 8 above)
Comment (aiyengar2) [reply cmd/kubectl.go]: I see. Is there any way we could use the structs defined in [`client-go/tools/clientcmd`](https://github.com/kubernetes/client-go/blob/30548acd0a9e231d42d406433aef55b111b36240/tools/clientcmd/api/v1/types.go#L28-L52) instead?

We could use [LoadFromFile](https://godoc.org/k8s.io/client-go/tools/clientcmd#LoadFromFile), [IsEmptyConfig](https://godoc.org/k8s.io/client-go/tools/clientcmd#IsEmptyConfig), and [WriteToFile](https://godoc.org/k8s.io/client-go/tools/clientcmd#WriteToFile). This might allow us to eliminate / simplify `loadKubeConfig`, `writeKubeConfig`, and `extractKubeconfigTokenID` here without requiring an additional inline struct.

File: cmd/server.go (hunk 3 above)
Comment (nickgerace) [reply cmd/server.go:73]: Ok sounds good

**Review by nickgerace — DISMISSED** [review]

File: cmd/kubectl.go (hunk 8 above)
Comment (rmweir) [reply cmd/kubectl.go]: sure, I can do- will need to add client-go as a dependency though.

File: cmd/kubectl.go [hunk 10]
```diff
@@ -74,3 +116,37 @@ func runKubectl(ctx *cli.Context) error {
...
+	return clientcmd.WriteToFile(*config, filePath)
+}
+
+func extractKubeconfigTokenID(kubeconfig api.Config) (string, error) {
+	if len(kubeconfig.AuthInfos) != 1 {
+		return "", fmt.Errorf("invalid kubeconfig, expected to contain exactly 1 user")
+	}
+	var parts []string
+	for _, val := range kubeconfig.AuthInfos {
```
Comment (dramich) [inline cmd/kubectl.go:133]: No need for the loop with the check for 1 above

File: cmd/kubectl.go [hunk 11]
```diff
@@ -37,34 +50,63 @@ func runKubectl(ctx *cli.Context) error {
...
 	if err != nil {
 		return err
 	}
-	defer os.Remove(tmpfile.Name())
 
-	_, err = tmpfile.Write([]byte(config.Config))
+	currentUser := t.UserID
+	currentCluster := currentRancherServer.FocusedCluster()
+	filename := fmt.Sprintf(kubeConfigNameFormat, currentUser, currentCluster)
+	kubeconfigFilePath := strings.Join([]string{kubeConfigPath, filename}, "/")
```
Comment (dramich) [inline cmd/kubectl.go:72]: Use path.Join here

File: cmd/kubectl.go [hunk 12]
```diff
@@ -2,11 +2,24 @@ package cmd
...
 	"fmt"
-	"io/ioutil"
 	"os"
 	"os/exec"
+	"strings"
 
 	"github.com/urfave/cli"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+
```
Comment (dramich) [inline cmd/kubectl.go:12]: Update the sorting here, delete the blank and sort again

File: cmd/server.go [hunk 13]
```diff
@@ -64,15 +64,13 @@ be displayed and one can be selected.
 }
 
 // serverCurrent command to display the name of the current server in the local config
-func serverCurrent(ctx *cli.Context) error {
-	cf, err := loadConfig(ctx)
+func printCurrentServer(ctx *cli.Context) error {
```
Comment (dramich) [inline cmd/server.go:67]: Comment should get updated as well

File: cmd/kubectl_token.go [hunk 14]
```diff
@@ -51,7 +51,7 @@ type LoginInput struct {
 }
 
 const (
-	kubeConfigCache = "/.cache/token"
+	tokenCache      = "/.cache/token"
```
Comment (dramich) [inline cmd/kubectl_token.go:54]: None of the changes in here are needed, this is getting changed in another pr

File: cmd/kubectl.go (hunk 10 above)
Comment (rmweir) [reply cmd/kubectl.go:133]: Spoke elsewhere, this is being used because the key is unknown.

File: cmd/kubectl.go [hunk 15]
```diff
@@ -37,29 +47,80 @@ func runKubectl(ctx *cli.Context) error {
...
 	}
 
-	cluster, err := getClusterByID(c, c.UserConfig.FocusedCluster())
+	config, err := loadConfig(ctx)
 	if err != nil {
 		return err
 	}
 
-	config, err := c.ManagementClient.Cluster.ActionGenerateKubeconfig(cluster)
+	currentRancherServer := config.FocusedServer()
```
Comment (nickgerace) [inline cmd/kubectl.go:55]: Should this function return `(server, err)` instead of just `server`? Might be a good change, but could be out of scope for this PR.

**Review by nickgerace — DISMISSED** [review]
LGTM barring one nit

File: config/config.go [hunk 16]
```diff
@@ -21,12 +22,13 @@ type Config struct {
...
-	URL       string `json:"url"`
-	Project   string `json:"project"`
-	CACerts   string `json:"cacert"`
+	AccessKey   string                 `json:"accessKey"`
+	SecretKey   string                 `json:"secretKey"`
+	TokenKey    string                 `json:"tokenKey"`
+	URL         string                 `json:"url"`
+	Project     string                 `json:"project"`
+	CACerts     string                 `json:"cacert"`
+	KubeConfigs map[string]*api.Config `json:"kubeConfigs"`
```
Comment (aiyengar2) [inline config/config.go:31]: nit: should this have the YAML fields as well or is the server config usually strictly JSON?

File: cmd/common.go [hunk 17]
```diff
@@ -139,6 +143,40 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
...
+func setKubeConfigForUser(ctx *cli.Context, user string, kubeConfig *api.Config) error {
+	cf, err := loadConfig(ctx)
+	if err != nil {
+		return err
+	}
+
+	if cf.FocusedServer().KubeConfigs == nil {
+		cf.FocusedServer().KubeConfigs = make(map[string]*api.Config)
+	}
+	cf.FocusedServer().KubeConfigs[fmt.Sprintf(kubeConfigKeyFormat, user, cf.FocusedServer().FocusedCluster())] = kubeConfig
```
Comment (aiyengar2) [inline cmd/common.go:176]: nit: should `cf.FocusedServer()` be just one call that gets stored in a variable?

File: cmd/common.go [hunk 18]
```diff
@@ -139,6 +143,40 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
...
+	return cf.CurrentServer, cf.Servers[serverName].URL, nil
+}
+
+func getCurrentKubeConfigForUser(ctx *cli.Context, user string) (*api.Config, error) {
+	cf, err := loadConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig, _ := cf.FocusedServer().KubeConfigs[fmt.Sprintf(kubeConfigKeyFormat, user, cf.FocusedServer().FocusedCluster())]
```
Comment (aiyengar2) [inline cmd/common.go:163]: nit: should `cf.FocusedServer()` be just one call that gets stored in a variable?

File: cmd/kubectl.go [hunk 19]
```diff
@@ -37,29 +47,80 @@ func runKubectl(ctx *cli.Context) error {
...
+		if err != nil {
+			return err
+		}
+		isTokenValid, err = validateToken(tokenID, c.ManagementClient.Token)
+		if err != nil {
+			return err
+		}
+	}
+
+	var kubeconfigBytes []byte
```
Comment (aiyengar2) [inline cmd/kubectl.go:84]: nit: `kubeconfigBytes` is inconsistent. Should be `kubeConfigBytes`

**Review by aiyengar2 — DISMISSED** [review]
Small comments, overall look good to me.

File: cmd/kubectl.go (hunk 15 above)
Comment (rmweir) [reply cmd/kubectl.go:51]: I don't think so, the function just accesses a map.

File: config/config.go (hunk 16 above)
Comment (rmweir) [reply config/config.go:31]: This is how it's been, I only added a field. I think most YAML libraries use the json tags now.

File: cmd/common.go (hunk 17 above)
Comment (rmweir) [reply cmd/common.go:176]: It's just access a map so it's not an expensive call, but it would probably be more readable, will change.

Comment (rmweir) [reply cmd/common.go:176]: Using a variable doesn't work for L174 since I want to assign a new values to one of it's fields.

**Review by aiyengar2 — DISMISSED** [review]
LGTM

**Review by nickgerace — DISMISSED** [review]

File: cmd/kubectl.go [hunk 20]
```diff
@@ -37,29 +43,80 @@ func runKubectl(ctx *cli.Context) error {
...
+			return err
+		}
+	}
+
+	if len(kubeConfigBytes) == 0 {
+		kubeConfigBytes, err = yaml.Marshal(kubeConfig)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (dramich) [inline cmd/kubectl.go:108]: kubeConfigBytes is not used after this so probably not needed

File: cmd/common.go [hunk 21]
```diff
@@ -139,6 +143,42 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
...
+	cf, err := loadConfig(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	serverName := cf.CurrentServer
+	return cf.CurrentServer, cf.Servers[serverName].URL, nil
+}
+
+func getCurrentKubeConfigForUser(ctx *cli.Context, user string) (*api.Config, error) {
```
Comment (dramich) [inline cmd/common.go:157]:
```suggestion
func getKubeConfigForUser(ctx *cli.Context, user string) (*api.Config, error) {
```

File: cmd/common.go [hunk 22]
```diff
@@ -139,6 +143,42 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
 	return writer.Err()
 }
 
+// getCurrentServer returns the name of the name and url of the current server in the local config
```
Comment (dramich) [inline cmd/common.go:146]:
```suggestion
// getCurrentServer returns the name and url of the current server in the local config
```

File: cmd/server.go [hunk 23]
```diff
@@ -63,16 +63,14 @@ be displayed and one can be selected.
...
+	serverName, url, err := getCurrentServer(ctx)
 	if err != nil {
 		return err
 	}
 
-	serverName := cf.CurrentServer
-	URL := cf.Servers[serverName].URL
-	fmt.Printf("Name: %s URL: %s\n", serverName, URL)
+	fmt.Printf("Name: %s URL: %s\n", serverName, url)
```
Comment (dramich) [inline cmd/server.go:73]: What is the purpose of this change? It doesn't seem related to the rest of the PR

File: cmd/common.go [hunk 24]
```diff
@@ -139,6 +143,42 @@ func listRoleTemplateBindings(ctx *cli.Context, b []RoleTemplateBinding) error {
...
+func getCurrentServer(ctx *cli.Context) (string, string, error) {
```
Comment (dramich) [inline cmd/common.go:147]: I don't think this is needed, it's only used in one place not related to the pr

**Review by dramich — CHANGES_REQUESTED** [review]

File: cmd/server.go (hunk 23 above)
Comment (rmweir) [reply cmd/server.go:73]: I set this up prior to moving to the config approach. I wanted to put it in common to be accessed elsewhere. No longer needed, will remove.

**Review by nickgerace — DISMISSED** [review]

File: cmd/kubectl.go [hunk 25]
```diff
@@ -37,29 +42,73 @@ func runKubectl(ctx *cli.Context) error {
...
+		if err != nil {
+			return err
+		}
+		isTokenValid, err = validateToken(tokenID, c.ManagementClient.Token)
+		if err != nil {
+			return err
+		}
+	}
+
+	var kubeConfigBytes []byte
```
Comment (dramich) [inline cmd/kubectl.go:79]: This is only used inside the if block so it probably doesn't need to be created here

**Review by dramich — APPROVED** [review]
